packages/typespec-ts/src/modular/emitSamples.ts

```diff
@@ -0,0 +1,516 @@
+  sourceFile.addStatements(`
+  async function main() {
+    ${functions}
+  }
+
+  main().catch(console.error);`);
+  options.generatedFiles.push(sourceFile);
+  return sourceFile;
+}
+
+function buildParameterValueMap(example: SdkHttpOperationExample) {
+  const parameterMap: Record<string, SdkHttpParameterExample> = {};
+  example.parameters.forEach(
+    (param) =>
+      (parameterMap[
+        (param.parameter as any).serializedName ?? param.parameter.name
+      ] = param)
+  );
+  return parameterMap;
+}
+
+function prepareMethodExampleParameters(
+  dpgContext: SdkContext,
+  exampleName: string,
+  method: SdkServiceMethod<SdkServiceOperation>,
+  parameterMap: Record<string, SdkHttpParameterExample>
+): ExampleValue[] {
+  const parameters: ExampleValue[] = [];
+  for (const param of method.operation.parameters) {
+    if (
+      param.optional === true ||
+      param.onClient === true ||
+      param.type.kind === "constant"
+    ) {
+      continue;
+    }
+    const exampleValue = parameterMap[param.serializedName];
+    if (!exampleValue || !exampleValue.value) {
+      // report diagnostic if required parameter is missing
+      reportDiagnostic(dpgContext.program, {
+        code: "required-sample-parameter",
+        format: {
+          exampleName: exampleName,
+          paramName: param.name
+        },
+        target: NoTarget
+      });
+      continue;
+    }
+
+    parameters.push({
+      name: param.name,
+      value: getParameterValue(exampleValue.value),
+      isOptional: false
+    });
+  }
+  // required/optional body parameters
+  const bodyParam = method.operation.bodyParam;
+  const bodyName = bodyParam?.name;
+  const bodyExample = parameterMap[bodyName ?? ""];
+  if (bodyName && bodyExample && bodyExample.value) {
+    if (
+      isSpreadBodyParameter(bodyParam) &&
+      bodyParam.type.kind === "model" &&
+      bodyExample.value.kind === "model"
+    ) {
+      for (const prop of bodyParam.type.properties) {
+        const propExample = bodyExample.value.value[prop.name];
+        if (!propExample) {
+          continue;
+        }
+        parameters.push({
+          name: prop.name,
+          value: getParameterValue(propExample),
+          isOptional: Boolean(prop.optional)
+        });
+      }
+    } else {
+      parameters.push({
+        name: bodyName,
+        value: getParameterValue(bodyExample.value),
+        isOptional: Boolean(method.operation.bodyParam?.optional)
+      });
+    }
+  }
+  // optional parameters
+  method.operation.parameters
+    .filter(
+      (param) => param.optional === true && parameterMap[param.serializedName]
+    )
+    .map((param) => parameterMap[param.serializedName]!)
+    .forEach((param) => {
+      parameters.push({
+        name: param.parameter.name,
+        value: getParameterValue(param.value),
+        isOptional: true
+      });
+    });
+  return parameters;
+}
+
+function prepareClientExampleParameters(
```

Comment:
why we need to differentiate how to prepare client parameters or method parameters? most of the logic seem similar.

Reply:
let me check if we could refactor this a little

Reply:
move to the same method.